applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp

```diff
@@ -23,25 +24,75 @@
 
 namespace Kratos {
 
+template<class TDataType>
 void ConversionUtilities::ConvertElementalDataToNodalData(
     ModelPart& rModelPart,
-    const Variable<array_1d<double,3> >& rElementalVariable,
-    const Variable<array_1d<double,3> >& rNodalVariable )
+    const Variable<TDataType>& rElementalVariable,
+    const Variable<TDataType>& rNodalVariable )
 {
     // prepare nodal variable
     VariableUtils().SetHistoricalVariableToZero(rNodalVariable, rModelPart.Nodes());
 
     block_for_each(rModelPart.Elements(), [&](Element& rElement){
-        const array_1d<double, 3>& elem_rVariable =  rElement.GetValue(rElementalVariable);
+        const auto& elem_rVariable =  rElement.GetValue(rElementalVariable);
 
         const std::size_t num_nodes = rElement.GetGeometry().PointsNumber();
 
         for (auto& r_node : rElement.GetGeometry().Points()){
-            AtomicAddVector( r_node.FastGetSolutionStepValue(rNodalVariable), (elem_rVariable / static_cast<double>(num_nodes)) );
+            if constexpr(std::is_same_v<TDataType, double>) {
+                AtomicAdd( r_node.FastGetSolutionStepValue(rNodalVariable), (elem_rVariable / static_cast<double>(num_nodes)) );
+            } else if constexpr(std::is_same_v<TDataType, array_1d<double, 3>>) {
+                AtomicAddVector( r_node.FastGetSolutionStepValue(rNodalVariable), (elem_rVariable / static_cast<double>(num_nodes)) );
+            } else {
+                static_assert(!std::is_same_v<TDataType, TDataType>, "Unsupported data type.");
+            }
         }
     });
 
     rModelPart.GetCommunicator().AssembleCurrentData(rNodalVariable);
 }
 
-}  // namespace Kratos.
+template<class TDataType>
+void ConversionUtilities::ConvertNodalDataToElementalData(
+    ModelPart& rModelPart,
+    const Variable<TDataType>& rElementalVariable,
+    const Variable<TDataType>& rNodalVariable )
+{
+    // prepare nodal variable
+    VariableUtils().SetHistoricalVariableToZero(rNodalVariable, rModelPart.Nodes());
```

Comment:
Why?

You're copying nonhistorical vars from nodes to elements, but why do you zero out the historical vars in the nodes?

Reply:
Good catch. If I set my nodal data to zero, how can my test pass?🤡

Reply:
Cuz you're zeroing out **historical** data but copying *non-historical* data.

Reply:
But in the test, I am using data location: nodal_historical ...

Reply:
> But in the test, I am using data location: nodal_historical ...

but you ignore it in your implementation :)

Reply:
I am trying to understand why does test passes and data with nodal_historical returns me correct values and not zeros if I use Get/SetValue in Python and cpp. Anyway, I will convert to use FastGetSolutilnValue and SetSolutionValue() and will see what happens.

Reply:
I think you're still mixing historical and non-historical storage.

- non-historical
    - `Node::SetValue`
    - `Node::GetValue`
    - slow access
    - allows different sets of vars for each entity in the list
    - doesn't segfault (querying a var that's not in the container inserts a null value)
    - cannot access values from previous time steps
    - usually used for material parameters, algorithmic parameters, and general stuff unrelated to FEM
- historical
    - `Node::GetSolutionStepValue` and `Node::FastGetSolutionStepValue`
    - nodes only
    - "fast" access
    - all nodes in the `ModelPart` must share the same set of vars
    - querying a var that's not in the container leads to
        - segfault in release mode (not a fan of this behavior)
        - exception in debug mode
    - stores values from a few recent time steps
    - used exclusively for variables we solve for (DoFs)

#
#
#

Looking at your test, it also mixes these up:

You're adding `TEMPERATURE` to the list of **historical** variables.
`applications/CoSimulationApplication/tests/test_coupling_operations.py:265`
```py
self.model_part.AddNodalSolutionStepVariable(KM.TEMPERATURE)
```

Then you're setting the **non-historical** `TEMPERATURE`.
`applications/CoSimulationApplication/tests/test_coupling_operations.py:270`
```py
new_node.SetValue(KM.TEMPERATURE, 5)
```

#
#
#

Then, you're zeroing out the **historical** `TEMPERATURE`s (which were already all zeros because you never set them).
`applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp:62`
```cpp
VariableUtils().SetVariable(rElementalVariable, rElementalVariable.Zero(), rModelPart.Nodes());
```
https://github.com/KratosMultiphysics/Kratos/blob/f8e5823de955e2579ddb5c6a0c5191d28081c5a0/kratos/utilities/variable_utils.h#L683-L685

Finally, you're copying the *non-historical* `TEMPERATURE` to the element, which you set in the test and never zeroed out.
`applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp:68-72`
```cpp
double temp = 0.0;
for (auto& r_node : rElement.GetGeometry().Points()){
    temp += r_node.GetValue(rNodalVariable) / num_nodes;
} 
rElement.SetValue(rElementalVariable, temp);
```

Reply:
@matekelemen well, now it is fixed=)